packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+        tagToAttributeMapping,
+        tagToBindAttributeMapping,
+      };
+    });
+  }
+
+  async transform(tree, params) {
+    if (!this.enabled) {
+      return;
+    }
+    const html = tree.root.firstChildByTag('html');
+    const head = html.firstChildByTag('head');
+    if (!head) return;
+
+    // Extensions which need to be imported
+    const extensionsToImport = new Set();
+    // Keep track of existing extensions imports to avoid duplicates
+    const existingImports = new Set();
+
+    // Some AMP components need to be detected in the head (e.g. amp-access)
+    await this.findExtensionsToImportInHead_(head, extensionsToImport, existingImports);
+
+    // Most AMP components can be detected in the body
+    const body = html.firstChildByTag('body');
+    await this.findExtensionsToImportInBody_(body, extensionsToImport);
+
+    if (extensionsToImport.length === 0) {
+      // Nothing to do
+      return;
+    }
+
+    // We use this for adding new import elements to the header
+    const referenceNode = findMetaViewport(head);
+
+    // Support custom runtime URLs
+    const host = calculateHost(params);
+    for (const extensionName of extensionsToImport) {
+      if (!existingImports.has(extensionName)) {
+        const extension = (await this.extensionSpec).extensionsMap.get(extensionName.trim());
+        this.log_.debug('auto importing', extensionName);
+        // Use the latest version by default
+        const version = extension.version[extension.version.length - 1];
+        const extensionImportAttribs = {
+          'async': '',
+          'src': `${host.ampUrlPrefix}/v0/${extensionName}-${version}.js`,
+        };
+        extensionImportAttribs[extension.type] = extensionName;
+        const extensionImport = tree.createElement('script', extensionImportAttribs);
+        head.insertAfter(extensionImport, referenceNode);
+      }
+    }
+  }
+
+  findExtensionsToImportInHead_(head, extensionsToImport, existingImports) {
+    let node = head.firstChild;
+    while (node) {
+      // Detect any existing extension imports
+      const customElement = this.getCustomElement_(node);
+      if (customElement) {
+        existingImports.add(customElement);
+      }
+      // Explicitly detect amp-access via the script tag in the header
+      if (node.tagName === 'script' && node.attribs['id'] === 'amp-access') {
+        extensionsToImport.add('amp-access');
+      }
+      node = node.nextSibling;
+    }
+  }
+
+  async findExtensionsToImportInBody_(body, extensionsToImport) {
+    let node = body;
+    const promises = [];
+    while (node !== null) {
+      if (node.tagName) {
+        promises.push(this.addRequiredExtensionByTag_(extensionsToImport, node));
+        promises.push(this.addRequiredExtensionByAttributes_(extensionsToImport, node));
+      }
+      node = node.nextNode();
+    }
+    return Promise.all(promises);
+  }
+
+  async addRequiredExtensionByTag_(allRequiredExtensions, node) {
```

Comment:
The argument order is reversed here (output set is first, while for `findExtensionsToImportInBody_` output set is last). I prefer first, but anything works as long as it's consistent.